### DISCUSSION TITLE
Missing python types for `ortools.linear_solver.pywraplp`

### DISCUSSION THREAD

**Discussion by grddavies** [google/or-tools#3186] [Linear Solver questions]
Issue:
-----
Missing type stubs for the `pywraplp` module.

Return types of various classes and methods are not compatible with static type checking with Pylance eg: `Solver.CreateSolver()`, `Solver.Objective()` . This causes type checking to fail when using pywraplp.

Request
-------
Request a .pyi file specifying python types for the classes and methods in pywraplp

Environment:
-------------
* Python v3.8.10
* ortools v9.3.10459
* Ubuntu 20.04 LTS
* VSCode with Pylance/Pyright

Errors:
------

Missing stub file warning:

>Stub file not found for "ortools.linear_solver" Pylance [reportMissingTypeStubs](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

Unknown Member Type Warning:
>Type of "CreateSolver" is partially unknown
 > Type of "CreateSolver" is "(solver_id: Unknown) -> Unknown" Pylance [reportUnknownMemberType](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)

**Comment by ghost** [comment]
Probably related to https://github.com/swig/swig/issues/735
A manual .pyi stub might be too much work

**Comment by lperron** [comment]
was it working before ?

**Comment by ghost** [comment]
I don't think so, this is probably more of a feature request

**Comment by lperron** [comment]
indeed.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 17 mars 2022 à 10:22, Xiang Chen ***@***.***> a
écrit :

**Comment by Mizux** [comment]
Since we are moving to pybind11 we should also check if it will be possible to do it with pybind11 tooling

**Reply by lperron** [reply]
We can check what typing is done on model_builder.

**Reply by Mizux** [reply]
Looking at the doc:
1. pybind11 provide most STL and plain type support
https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html#conversion-table
2. Some related stuff, it seems pybind11 can/could generate clean docstring signature but don't know if IDE can use it like regular typing in a .py file.
https://pybind11.readthedocs.io/en/stable/advanced/misc.html#avoiding-c-types-in-docstrings
https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods

**Reply by ghost** [reply]
@Mizux I tried [stubgen](https://mypy.readthedocs.io/en/stable/stubgen.html) with ` stubgen -p ortools` and it did a pretty great job with both swig and pybind11 generated files.

If you go with this route do make sure to not replace the `.pyi` files from `mypy-protobuf` though, as this tool doesn't work that well with `_pb2.py` files.

**Reply by Mizux** [reply]
Seems stugen support module directory or filename so:
module without proto -> stubgen module dir, otherwise file per file `-_-`
